core/src/main/java/org/infinispan/commands/irac/IracCleanupTombstoneCommand.java

```diff
@@ -0,0 +1,149 @@
+      super(COMMAND_ID, cacheName);
+      this.key = key;
+      this.tombstone = tombstone;
+   }
+
+   @Override
+   public ByteString getCacheName() {
+      return cacheName;
+   }
+
+   @Override
+   public CompletionStage<Boolean> invokeAsync(ComponentRegistry registry) {
+      if (tombstone == null) {
+         // command received from a remote site.
+         // check if the key exists in IracManager
+         return isKeyInIracManager(registry);
+      }
+      // removes the tombstone
+      registry.getIracTombstoneCleaner().running().removeTombstone(key, tombstone);
+      return CompletableFutures.completedNull();
+   }
+
+   @Override
+   public byte getCommandId() {
+      return COMMAND_ID;
+   }
+
+   @Override
+   public CompletionStage<Boolean> performInLocalSite(ComponentRegistry registry, boolean preserveOrder) {
+      DistributionInfo distribution = registry.getDistributionManager().getCacheTopology().getDistribution(key);
+      if (distribution.isPrimary()) {
+         return isKeyInIracManager(registry);
+      } else {
+         RpcManager manager = registry.getRpcManager().running();
+         return manager.invokeCommand(distribution.primary(), this, new BooleanResponseCollector(), manager.getSyncRpcOptions());
+      }
+   }
+
+   @Override
+   public CompletionStage<Boolean> performInLocalSite(BackupReceiver receiver, boolean preserveOrder) {
+      throw new IllegalStateException("Should never be invoked!");
+   }
+
+   @Override
+   public boolean isReturnValueExpected() {
+      return true;
+   }
+
+   @Override
+   public void writeTo(ObjectOutput output) throws IOException {
+      output.writeObject(key);
+      IracMetadata.writeTo(output, tombstone);
+   }
+
+   @Override
+   public void readFrom(ObjectInput input) throws IOException, ClassNotFoundException {
+      this.key = input.readObject();
+      this.tombstone = IracMetadata.readFrom(input);
+   }
+
+   @Override
+   public Address getOrigin() {
+      //not needed
+      return null;
+   }
+
+   @Override
+   public void setOrigin(Address origin) {
+      //no-op
+   }
+
+   @Override
+   public String toString() {
+      return "IracCleanupTombstoneCommand{" +
+            "cacheName=" + cacheName +
+            ", key=" + Util.toStr(key) +
+            ", tombstone=" + tombstone +
+            '}';
+   }
+
+   private CompletionStage<Boolean> isKeyInIracManager(ComponentRegistry registry) {
+      return CompletableFutures.booleanStage(registry.getIracManager().running().containsKey(key));
+   }
+
+   private static final class BooleanResponseCollector extends ValidSingleResponseCollector<Boolean> {
+
+      @Override
+      protected Boolean withValidResponse(Address sender, ValidResponse response) {
+         return (Boolean) response.getResponseValue();
+      }
+
+      @Override
+      protected Boolean targetNotFound(Address sender) {
+         return Boolean.TRUE;
```

Comment:
Wouldn't it be cheaper if the site coordinator waited for a new topology and retried, instead of assuming an update is ongoing and doing another xsite RPC?

Reply:
yes, I think so. It is worth implementing it now? 
I mean, the probability of a request arriving while the primary owner is leaving should be quite low IMO.
WDYT? Implement it now or create a JIRA for later?

Reply:
Create a JIRA for later, sure.